File: algorithms/11_golang/selection/quickselect.go
package main

import (
	"fmt"
	"sort"
)

func quickselect(a []int, k int) int {
	lo, hi := 0, len(a)-1
	for {
		pivot, i := a[hi], lo
		for j := lo; j < hi; j++ {
			if a[j] < pivot {
				a[i], a[j] = a[j], a[i]
				i++
			}
		}
		a[i], a[hi] = a[hi], a[i]
		if i == k {
			return a[i]
		}
		if k < i {
			hi = i - 1
		} else {
			lo = i + 1
		}
	}
}

func main() {
	data := []int{3, 2, 1, 5, 6, 4}
	sorted := append([]int(nil), data...)
	sort.Ints(sorted)
	for k := range data {
		got := quickselect(append([]int(nil), data...), k)
		if got != sorted[k] {
			panic("quickselect mismatch")
		}
	}
	fmt.Println("quickselect ok")
}

File: algorithms/06_javascript/selection/quickselect.js
function quickselect(values, k) {
  const a = [...values];
  let lo = 0, hi = a.length - 1;
  while (true) {
    const pivot = a[hi];
    let i = lo;
    for (let j = lo; j < hi; j += 1) {
      if (a[j] < pivot) {
        [a[i], a[j]] = [a[j], a[i]];
        i += 1;
      }
    }
    [a[i], a[hi]] = [a[hi], a[i]];
    if (i === k) return a[i];
    if (k < i) hi = i - 1; else lo = i + 1;
  }
}

module.exports = { quickselect };

if (require.main === module) {
  const data = [3, 2, 1, 5, 6, 4];
  const sorted = [...data].sort((x, y) => x - y);
  for (let k = 0; k < data.length; k += 1) {
    if (quickselect(data, k) !== sorted[k]) throw new Error("quickselect failed");
  }
  console.log("[JavaScript Quickselect] order statistics verified");
}
